File: guests/go/richtext/richtext.go
// The rich text scene (tools/scenes/richtext.steps): the app declares a
// document, applies an edit, formats the widget's selection through its
// own act, and reads every delta back into its own Document. THE OFFSETS
// ARE UTF-8 BYTES, which is what Go's own string indexing already
// produces (docs/ranges-units.md).
package richtext

import (
	"fmt"
	"strings"

	kaya "dev.kaya/bindings/go"
)

const document = "Héllo world\nSecond line"

// spell is the core's spelling of runs (`expect_runs`), so the binding's
// document and the core's mirror are compared as one string.
func spell(runs []kaya.TextRun) string {
	parts := make([]string, 0, len(runs))
	for _, run := range runs {
		if run.Value == "true" {
			parts = append(parts, fmt.Sprintf("%d:%d %s", run.Start, run.End, run.Name))
			continue
		}
		parts = append(parts, fmt.Sprintf("%d:%d %s=%s", run.Start, run.End, run.Name, run.Value))
	}
	return strings.Join(parts, "|")
}

func App() *kaya.App {
	app := kaya.NewApp()

	var last, runs kaya.Signal[string]
	var editor kaya.Widget

	app.Build(func(tx *kaya.Tx) {
		tx.Window(0).Title("richtext")
		last = tx.Signal("")
		runs = tx.Signal("")

		tx.Mount(tx.Column(func() {
			editor = tx.Textarea(nil).Rich().A11yID("doc").A11yLabel("Document")
			app.OnEdit(editor, func(tx *kaya.Tx, edit kaya.Edit) {
				tx.Write(last, fmt.Sprintf("edit %d:%d <%s> [%s]",
					edit.Start, edit.End, edit.Inserted, spell(edit.Runs)))
				tx.Write(runs, spell(app.Document(editor).Runs))
			})
			app.OnFormat(editor, func(tx *kaya.Tx, act kaya.Format) {
				value := act.Value
				if act.Removed {
					value = "off"
				}
				tx.Write(last, fmt.Sprintf("format %d:%d %s=%s",
					act.Start, act.End, act.Name, value))
				tx.Write(runs, spell(app.Document(editor).Runs))
			})

			tx.Label(last) // label#0
			tx.Label(runs) // label#1

			tx.Row(func() {
				tx.Button("seed", func(tx *kaya.Tx) { // button#0
					doc := kaya.NewDocument(document).
						Bold(0, 6).
						Link(7, 12, "https://kaya.dev").
						Block(13, 24, kaya.Heading2)
					tx.SetDocument(editor, doc)
					tx.Write(runs, spell(doc.Runs))
				})
				tx.Button("insert", func(tx *kaya.Tx) { // button#1
					tx.ApplyEdit(editor, kaya.Insert(6, ", big").Mark(2, 5, "italic", "true"))
					tx.Write(runs, spell(app.Document(editor).Runs))
				})
				tx.Button("select word", func(tx *kaya.Tx) { // button#2
					tx.SelectRange(editor, kaya.TextRange{Start: 0, End: 6})
				})
				tx.Button("unbold", func(tx *kaya.Tx) { // button#3
					tx.Unformat(editor, "bold")
				})
				tx.Button("heading", func(tx *kaya.Tx) { // button#4
					tx.SetBlock(editor, kaya.Heading1)
				})
				tx.Button("focus", func(tx *kaya.Tx) { // button#5
					tx.Focus(editor)
				})
				tx.Button("prefix", func(tx *kaya.Tx) { // button#6
					tx.ApplyEdit(editor, kaya.Insert(0, "> "))
					tx.Write(runs, spell(app.Document(editor).Runs))
				})
			})
		}))
	})

	return app
}

File: guests/java/dev/kaya/guests/RichText.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

import java.util.ArrayList;
import java.util.List;

/**
 * The rich text scene from the JVM — guests/rust/richtext.rs,
 * tools/scenes/richtext.steps. THE OFFSETS ARE UTF-8 BYTES; the
 * accented first word is what makes a UTF-16 reader fail
 * (docs/ranges-units.md), so every range here goes through
 * TextRange.in, which converts against the text it indexes.
 */
public final class RichText {
    /** Java lambdas cannot assign captured locals. */
    /** A SELF-REFERENCE, not a smuggle: the textarea's own onEdit reads
     * the document of the widget being declared. */
    private static final class Refs {
        KayaApp.Widget editor;
    }

    /** THE ACCENT IS IN A {@code \}{@code u} ESCAPE: a literal's bytes
     * follow the compiler's encoding, and the Android build that also
     * compiles this directory passes no {@code -encoding}
     * (android/javahost/build.gradle.kts). */
    private static final String DOC = "H\u00e9llo world\nSecond line";

    private static final String URL = "https://kaya.dev";

    /** The core's spelling of runs ({@code expect_runs}), so the
     * binding's document and the core's mirror are compared as one
     * string. */
    private static String spell(List<KayaApp.TextRun> runs) {
        List<String> parts = new ArrayList<>();
        for (KayaApp.TextRun run : runs) {
            parts.add(run.isFlag()
                    ? run.range().start + ":" + run.range().stop + " " + run.name()
                    : run.range().start + ":" + run.range().stop + " " + run.name() + "=" + run.value());
        }
        return String.join("|", parts);
    }

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            tx.window(0).title("richtext");
            KayaApp.Signal<String> last = tx.signal("");
            KayaApp.Signal<String> runs = tx.signal("");

            Refs refs = new Refs();
            tx.mount(tx.column(col -> {
                refs.editor = tx.textarea().rich().a11yId("doc").a11yLabel("Document");
                app.onEdit(refs.editor, (t, edit) -> {
                    String source = edit.source() == null ? "?" : edit.source().toString();
                    t.write(last, "edit " + edit.range().start + ":" + edit.range().stop + " <"
                            + edit.inserted() + "> " + source + " ["
                            + spell(edit.runs()) + "]");
                    t.write(runs, spell(app.document(refs.editor).runs()));
                });
                app.onFormat(refs.editor, (t, act) -> {
                    t.write(last, "format " + act.range().start + ":" + act.range().stop + " "
                            + act.name() + "=" + (act.value() == null ? "off" : act.value()));
                    t.write(runs, spell(app.document(refs.editor).runs()));
                });

                tx.label(last); // label#0
                tx.label(runs); // label#1

                tx.row(row -> {
                    tx.button("seed", t -> { // button#0
                        KayaApp.Document doc = new KayaApp.Document(DOC)
                                .bold(KayaApp.TextRange.in(DOC, 0, 5))
                                .link(KayaApp.TextRange.in(DOC, 6, 11), URL)
                                .block(KayaApp.TextRange.in(DOC, 12, 23),
                                        KayaApp.Block.HEADING2);
                        t.setDocument(refs.editor, doc);
                        t.write(runs, spell(doc.runs()));
                    });
                    tx.button("insert", t -> { // button#1
                        KayaApp.Edit edit =
                                KayaApp.Edit.insert(KayaApp.TextRange.in(DOC, 5, 5), ", big")
                                        .mark(KayaApp.TextRange.in(", big", 2, 5), "italic",
                                                true);
                        t.applyEdit(refs.editor, edit);
                        t.write(runs, spell(app.document(refs.editor).runs()));
                    });
                    tx.button("select word", t -> // button#2
                            t.selectRange(refs.editor, KayaApp.TextRange.in(DOC, 0, 5)));
                    tx.button("unbold", t -> t.unformat(refs.editor, "bold")); // button#3
                    tx.button("heading", t -> // button#4
                            t.setBlock(refs.editor, KayaApp.Block.HEADING1));
                    tx.button("focus", t -> t.focus(refs.editor)); // button#5
                    tx.button("prefix", t -> { // button#6
                        t.applyEdit(refs.editor,
                                KayaApp.Edit.insert(KayaApp.TextRange.in(DOC, 0, 0), "> "));
                        t.write(runs, spell(app.document(refs.editor).runs()));
                    });
                });
            }));
        });

        app.dispatchLoop();
    }

    private RichText() {}
}
